### PR TITLE
Add support for optional connections

### LINKED ISSUE #3703: Support for nullable Connection
The field using `strawberry.connection` should be able to be nullable.

## Feature Request Type

- [ ] Core functionality
- [x] Alteration (enhancement/optimization) of existing feature(s)
- [x] New behavior

## Description

This is needed in case if connection field would return an error e.g. from `PermissionExtension` and we don't want to error out the whole query.
Right now when the field has a declared type of `strawberry.relay.ListConnection[...] | None` the following error is thrown
```
strawberry.relay.exceptions.RelayWrongAnnotationError: Wrong annotation used on field "...". It should be annotated with a "Connection" subclass.
```

### PR DESCRIPTION
**Pull request by patrick91**
Closes #3703

## Summary by Sourcery

Add support for optional Relay connections to enhance flexibility in permission handling and update tests and documentation accordingly.

New Features:
- Add support for making Relay connections optional, allowing for more flexible permission handling.

Documentation:
- Add a new release note detailing the introduction of optional Relay connections and providing an example of its usage.

Tests:
- Introduce tests to verify the functionality of optional Relay connections, including scenarios with permissions and different Python versions.

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release adds support for making Relay connection optional, this is useful
when you want to add permission classes to the connection and not fail the whole
query if the user doesn't have permission to access the connection.

Example:

```python
import strawberry
from strawberry import relay
from strawberry.permission import BasePermission


class IsAuthenticated(BasePermission):
    message = "User is not authenticated"

    # This method can also be async!
    def has_permission(
        self, source: typing.Any, info: strawberry.Info, **kwargs
    ) -> bool:
        return False


@strawberry.type
class Fruit(relay.Node):
    code: relay.NodeID[int]
    name: str
    weight: float

    @classmethod
    def resolve_nodes(
        cls,
        *,
        info: strawberry.Info,
        node_ids: Iterable[str],
    ):
        return []


@strawberry.type
class Query:
    node: relay.Node = relay.node()

    @relay.connection(
        relay.ListConnection[Fruit] | None, permission_classes=[IsAuthenticated()]
    )
    def fruits(self) -> Iterable[Fruit]:
        # This can be a database query, a generator, an async generator, etc
        return all_fruits.values()
```



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

File: strawberry/relay/fields.py [hunk 1]
```diff
@@ -253,13 +256,18 @@ def apply(self, field: StrawberryField) -> None:
                 None,
             )
 
+        if is_union(resolver_type):
+            # TODO: actually check if is optional and get correct type
+            resolver_type = get_args(resolver_type)[0]
```
Comment (bellini666) [inline strawberry/relay/fields.py:261]: **suggestion:** we can probably use `is_optional` to check and `get_optional_annotation` in this line from `strawberry.utils.typing`

**Review by bellini666 — APPROVED** [review]
LGTM :)

Left a couple of suggestions

File: strawberry/schema/subscribe.py [hunk 2]
```diff
@@ -138,7 +138,7 @@ async def subscribe(
         middleware_manager,
         execution_context_class,
     )
-    # GrapQL-core might return an initial error result instead of an async iterator.
+    # GraphQL-core might return an initial error result instead of an async iterator.
```
[the remark this answers is not in the document]
Comment (bellini666) [reply strawberry/schema/subscribe.py:141]: Hrm, isn't this suggestion correct? 🤔

**Pull request merged**